proto/public/uber/cadence/api/v1/decision.proto

```diff
@@ -0,0 +1,171 @@
+  ActivityType activity_type = 2;
+  string domain = 3;
+  TaskList task_list = 4;
+  Payload input = 5;
+  google.protobuf.Duration schedule_to_close_timeout = 6 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration schedule_to_start_timeout = 7 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration start_to_close_timeout = 8 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration heartbeat_timeout = 9 [(gogoproto.stdduration) = true];
+  RetryPolicy retry_policy = 11;
+  Header header = 12;
+  bool request_local_dispatch = 13;
+}
+
+message StartTimerDecisionAttributes {
+  string timer_id = 1;
+  google.protobuf.Duration start_to_fire_timeout = 2 [(gogoproto.stdduration) = true];
+}
+
+message CompleteWorkflowExecutionDecisionAttributes {
+  Payload result = 1;
+}
+
+message FailWorkflowExecutionDecisionAttributes {
+  Failure failure = 1;
+}
+
+message RequestCancelActivityTaskDecisionAttributes {
+  string activity_id = 1;
+}
+
+message CancelTimerDecisionAttributes {
+  string timer_id = 1;
+}
+
+message CancelWorkflowExecutionDecisionAttributes {
+  Payload details = 1;
+}
+
+message RequestCancelExternalWorkflowExecutionDecisionAttributes {
+  string domain = 1;
+  WorkflowExecution workflow_execution = 2;
+  bytes control = 3;
+  bool child_workflow_only = 4;
+}
+
+message RecordMarkerDecisionAttributes {
+  string marker_name = 1;
+  Payload details = 2;
+  Header header = 3;
+}
+
+message ContinueAsNewWorkflowExecutionDecisionAttributes {
+  WorkflowType workflow_type = 1;
+  TaskList task_list = 2;
+  Payload input = 3;
+  google.protobuf.Duration execution_start_to_close_timeout = 4 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration task_start_to_close_timeout = 5 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration backoff_start_interval = 6 [(gogoproto.stdduration) = true];
+  RetryPolicy retry_policy = 7;
+  ContinueAsNewInitiator initiator = 8;
+  Failure failure = 9;
+  Payload last_completion_result = 10;
+  string cron_schedule = 11;
+  Header header = 12;
+  Memo memo = 13;
+  SearchAttributes search_attributes = 14;
+}
+
+message StartChildWorkflowExecutionDecisionAttributes {
+  string domain = 1;
+  string workflow_id = 2;
+  WorkflowType workflow_type = 3;
+  TaskList task_list = 4;
+  Payload input = 5;
+  google.protobuf.Duration execution_start_to_close_timeout = 6 [(gogoproto.stdduration) = true];
+  google.protobuf.Duration task_start_to_close_timeout = 7 [(gogoproto.stdduration) = true];
+  ParentClosePolicy parent_close_policy = 8;
+  bytes control = 9;
+  WorkflowIdReusePolicy workflow_id_reuse_policy = 10;
+  RetryPolicy retry_policy = 11;
+  string cron_schedule = 12;
+  Header header = 13;
+  Memo memo = 14;
+  SearchAttributes search_attributes = 15;
+}
+
+message SignalExternalWorkflowExecutionDecisionAttributes {
+  string domain = 1;
+  WorkflowExecution workflow_execution = 2;
+  string signal_name = 3;
+  Payload input = 4;
+  bytes control = 5;
```

Comment:
what's "control" for?

Reply:
TBH, I'm not entirely sure about their purpose. They seem to go all way along the commit history and allows some arbitrary data passed along when dealing with external or child workflows. Couple of places that actually uses them: [here](https://github.com/uber-go/cadence-client/blob/4653b9fd2a2ee0b7b9b6cfd5245b3f5f0414e325/internal/internal_decision_state_machine.go#L878) and [here](https://github.com/uber-go/cadence-client/blob/4653b9fd2a2ee0b7b9b6cfd5245b3f5f0414e325/internal/internal_decision_state_machine.go#L938). Both of them are strings converted to bytes.

I'm not aware about the origin intention of these fields, therefore it is hard to reason about them. Currently they are the same as in Thrift.

Reply:
heh.  good enough for me at least :)  maybe I'll understand them after reading more of the source.